mfront-doc/include/MFront/DocumentationGeneratorBase.hxx

```diff
@@ -0,0 +1,117 @@
+/*!
+ * \file  DocumentationGeneratorBase.hxx
+ * \brief
+ * \author Maxence Wangermez
+ * \date   16 avril 2024
+ * \copyright Copyright (C) 2006-2024 CEA/DEN, EDF R&D. All rights
+ * reserved.
+ * This project is publicly released under either the GNU GPL Licence
+ * or the CECILL-A licence. A copy of thoses licences are delivered
+ * with the sources of TFEL. CEA or EDF may also distribute this
+ * project under specific licensing conditions.
+ */
+
+#ifndef LIB_MFRONT_DOCUMENTATIONGENERATORBASE_H
+#define LIB_MFRONT_DOCUMENTATIONGENERATORBASE_H
+
+#include <string>
+#include <vector>
+#include <memory>
+#include <utility>
+#include <functional>
+
+#include "TFEL/Utilities/ArgumentParserBase.hxx"
+#include "TFEL/Material/ModellingHypothesis.hxx"
+#include "MFront/MFrontBase.hxx"
+
+namespace mfront {
+
+  // forward declaration
+  struct BehaviourDescription;
+  // forward declaration
+  struct FileDescription;
+
+  /*!
+   * Class used by the mfront-query tool to extract information from
+   * behaviour implementation
+   */
+  struct DocumentationGeneratorBase
+      : public tfel::utilities::ArgumentParserBase<DocumentationGeneratorBase>,
+        public MFrontBase {
+    //! \brief type of documentation to be generated
+    enum OutputType { FULL, WEB };
+    /*!
+     * build a DocumentationGeneratorBase object based on command line
+     * arguments
+     * \param[in] argc : number of command line arguments
+     * \param[in] argv : command line arguments
+     * \param[in] f    : filename
+     */
+    DocumentationGeneratorBase(const int,
+                               const char *const *const,
+                               const std::string &);
+    virtual void exe() const = 0;
+    //! destructor
+    ~DocumentationGeneratorBase() override;
+
+    // Private member data
+    //! input file name
+    std::string file;
+    /*!
+     * \brief boolean whose meaning is the following:
+     * - if true, the generated documentation is contained in a whole document.
+     * - if false, the generated documentation is contained in a section meant
+     *   to be included in a bigger document.
+     */
+    bool standalone = false;
+
+    //! \brief generate output on standard output
+    bool std_output = false;
+    //! type of ouput
+    OutputType otype;
+
+    /*!
+     * internal structure gathering data from mechanical behaviour
+     * description
+     */
+    struct Data {
+      using Hypothesis = tfel::material::ModellingHypothesis::Hypothesis;
+      Data();
+      Data(Data &&) noexcept;
+      Data(const Data &);
+      Data &operator=(Data &&);
+      Data &operator=(const Data &);
+      ~Data() noexcept;
+      std::string name;
+      std::string type;
+      std::string description;
+      //! specific description per modelling hypothesis
+      std::map<Hypothesis, std::string> descriptions;
+      std::string externalName;
+      std::vector<Hypothesis> hypotheses;
+      unsigned short arraySize;
+    };
```

Comment:
The name of this class is not appropriate, but that's my fault :)

However, it is not used in `DocumentationGeneratorBase` and it contains some members only useful for behaviours.

```suggestion
    //! \brief data structure describing a variable
    struct VariableInformationBase {
      VariableInformationBase();
      VariableInformationBase(VariableInformationBase &&) noexcept;
      VariableInformationBase(const VariableInformationBase &);
      VariableInformationBase &operator=(VariableInformationBase &&);
      VariableInformationBase &operator=(const VariableInformationBase &);
      ~VariableInformationBase() noexcept;
      std::string name;
      std::string type;
      std::string description;
      std::string externalName;
      unsigned short arraySize;
    };
```

and create two other data structures both named `VariableInformation` in `MaterialPropertyDocumentationGenerator` and `BehaviourDocumentationGenerator`.

In `MaterialPropertyDocumentationGenerator`:

```c++
using VariableInterformation = VariableInformationBase;
```

In `BehaviourDocumentationGenerator`:

```c++
struct VariableInterformation : VariableInformationBase{
      VariableInformation();
      VariableInformation(VariableInformation &&) noexcept;
      VariableInformation(const VariableInformation &);
      VariableInformation &operator=(VariableInformation &&);
      VariableInformation &operator=(const VariableInformation &);
      ~VariableInformation() noexcept;
      std::map<Hypothesis, std::string> descriptions;
      std::vector<Hypothesis> hypotheses;
};
```